### PR TITLE
unix: delay signal handling until after normal i/o

### PR DESCRIPTION
**Pull request by bnoordhuis**
It was reported that some node.js tests fail on AIX because the exit
event sometimes comes before the final stdio output of a child process.

Work around that by deferring the signal watcher that is used for
process management until after the dispatch of regular i/o watchers.

Fixes: https://github.com/libuv/libuv/issues/610

R=@saghul.  The duplication is galling, we should do something about that in v2.

[I'll update with a CI link once ci.nodejs.org is back up.]

### CODE REVIEW

**Comment by trevnorris** [comment]
Just curious, is it not possible to guarantee the same on win?

**Comment by bnoordhuis** [comment]
Probably, although I'm not sure what the proper way to do it is.  The best I can come up with is run `UV_PROCESS_EXIT` and `UV_SIGNAL_REQ` requests last in `uv_process_reqs()`.

**Comment by piscisaureus** [comment]
I believe this issue is fundamentally un-fixable; imagine this scenario:

* A libuv process ("master") spawns a child process ("intermediate")
* The "intermediate" child process spawns another process ("leaf") which inherits the stdout file descriptor.
* The "intermediate" process exits.

Now the parent will receive an exit notification, however the associated stdout pipe may remain open indefinitely because the "leaf" process is still using it.

In node we fixed this by creating two separate event (`exit` and `close`), where the `exit` event may fire before stdio is flushed and `close` is postponed until all stdio pipes are closed - which may never happen!

**Comment by bnoordhuis** [comment]
I don't disagree, I labeled it a quality-of-implementation issue in the node.js bug report.  This PR makes the common case behave like people expect it to, it doesn't (can't) fix all cases.

**Comment by saghul** [comment]
CI: https://ci.nodejs.org/job/libuv+any-pr+multi/185/

**Comment by saghul** [comment]
LGTM. Hopefully we can DRY this out for v2.

**Comment by saghul** [comment]
The CI is happy, BTW. (for out definition of happy, that is)

**Comment by bnoordhuis** [comment]
@piscisaureus You want to comment on whether this should be done for the Windows side too?

**Comment by saghul** [comment]
Ping?

**Comment by bnoordhuis** [comment]
More precisely: ping @piscisaureus.

**Comment by gireeshpunathil** [comment]
A gentle ping to know the status.

**Comment by mhdawson** [comment]
Can this land we still see intermittent failures in our testing on AIX due to this.

**Comment by bnoordhuis** [comment]
Rebased and new CI: https://ci.nodejs.org/job/libuv-test-commit/32/

I'll give it one more day.  @piscisaureus If you still want to chime in?

**Comment by saghul** [comment]
I'll defer this one to @bnoordhuis

**Comment by piscisaureus** [comment]
Sorry for not responding ... on Windows we could prioritize pipe i/o over the child process exit event, but I'm not even sure that this issue exists on windows. It seems inherently racy to me (even after this fix) and on windows the pipe i/o may end up going first anyway.

**Comment by mhdawson** [comment]
@piscisaureus  can't tell if you are suggesting some additional changes or the existing PR is ok as is as you don't think there is an issue on windows or we need do something else ?

**Comment by piscisaureus** [comment]
@mhdawson Woops, I dropped the quote that I was responding too; put it back now.

I was trying to answer the question whether windows needs a similar fix. I'm saying that windows is affected by the same problem (at least in theory), but I don't think there's merit to making changes, because:
  - no windows users are reporting this problem
  - a deterministically correct solution for this issue doesn't exist.

So long story short, we're done here.
